### ISSUE TITLE
Windows DirectX VideoPlayer leaks large amounts of memory

### ISSUE THREAD

**Issue by jbeshir** [MonoGame/MonoGame#4121]
Each WMV video played would leak about 75MB of memory, rapidly increasing the amount used by the software, with disposing the videos doing nothing, even after PlatformDispose was uncommented in Video.cs.

I spent a while looking into it and came to the conclusion that the cleanup process wasn't being handled correctly; MediaFoundation documentation at https://msdn.microsoft.com/en-us/library/windows/desktop/ms703190(v=vs.85).aspx describes an asynchronous process of closing the session that needs to be done before the media source, then session is disposed. I couldn't find any documentation for disposing media sources without closing the session they had been used with.

I came up with a first draft for a set of changes to fix this, using #4090's work on MediaPlayer as a starting point, which move the MediaSession into the Video class and have a MediaSession per video, which is closed, followed by disposal of the rest of the resources, when the video is disposed. They also dispose events received by our callback class, since not doing so causes the leak to persist, which is something that likely wants to be incorporated in any fixes to MediaPlayer's disposing of resources, too. The changes are at https://github.com/jbeshir/MonoGame/tree/video-player-rework but this isn't suitable to merge because of some remaining issues I encountered while working on it:
- The TaskScheduler-based dispatching to the UI thread used in #4090 didn't seem to be working right for me, it was running things on the thread pool, so I had to incorporate the same approach to passing tasks off to the UI thread used on Android. I suspect this could have been caused in some fashion by the particular product I was testing with, though.
- Because it needs to invoke to the UI thread several times to raise events before the video enters the playing state after a call to PlatformPlay, the loop waiting for the session to enter the playing state in Play always fails. I'm not sure the best way to deal with this, but this might not be a problem if the above is fixed since it sleeps in Play and .NET pumps the message queue while sleeping/blocked on a lock by default.

(In our own fork I worked around this by adding a PlayAsync method to VideoPlayer.cs, which dropped the loop-and-sleep-until-state-changed behaviour entirely, since it was desirable to make things smoother anyway by letting the rest of the program continue while the video setup happened, but this didn't fix the XNA compatible API and I've not included it in the linked branch.)

**Comment by tomspilman** [comment]
We found the same thing in an internal project.  I haven't had time to get our code changes wrapped up and into a PR yet.

**Comment by mrhelmut** [comment]
I just noticed that the task scheduler problem you have also occurs in MediaPlayer (in PR #4090).
If the changes in #4090 prove themselves to be stable, they can definitely be applied to VideoPlayer!

**Comment by Jjagg** [comment]
Closing as duplicate of #5308.
